File: C++/8.cpp
class Solution {
public:
    int myAtoi(string str) {
        long long const MAX = 2147483648LL;
        auto it = str.begin();
        bool neg = false;
        long long value = 0;
        /*
            首先去除开始的空格
        */
        while(it != str.end() && *it == ' ') ++it;
        /*
            处理可能存在的符号位
        */
        if(it != str.end() && (*it == '+' || *it == '-'))
        {
            if(*it == '-') neg = true;
            ++it;
        }
        /*
            处理数字，直到第一个不是数字的字符出现
        */
        while(it != str.end())
        {
            if(!isdigit(*it)) break;
            value *= 10;
            value += *it - '0';
            ++it;
            if(value > MAX) break;
        }
        /*
            处理数据溢出的情况
        */
        if(neg)
        {
            if(value > MAX) return -MAX;
            else return -value;
        }
        else
        {
            if(value > MAX - 1) return MAX - 1;
            else return value;
        }
    }
};

File: Python/8.py
#!/usr/bin/env python
# encoding: utf-8

class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        '''
        基本使用内置: 居然超过了100%...可能判断机制做了修改
        '''
        s=re.search(r'^([+-]?\d+)',str.strip())
        if not s:
            return 0
        res = int(s.group(1))
        if res>0 and abs(res) > 2147483647:
            return 2147483647
        if res<0 and abs(res)> 2147483648:
            return -2147483648
        return res
